### ISSUE TITLE
New release process?

### ISSUE THREAD

**Issue by theacodes** [urllib3/urllib3#1392]
Having just gone through the release process the first time myself I found that I wasn't a huge fan of it.

I propose we do the following for releases, but I can be convinced otherwise:

1. Create a new branch, `release-MAJOR.MINOR.PATCH`.
1. Add a single commit to that branch that updates the version in `urllib3/__init__.py` and `CHANGES.rst`.
1. Send a PR to merge that branch into `master`. Someone *must* review.
1. Merge the PR and delete the `HEAD` branch.
1. Tag the *squashed merge commit* with `MAJOR.MINOR.PATCH`.
1. Let Travis upload the package to PyPI when built with tags.

This would mean *removing* the `release` branch.

/cc @urllib3/maintainers for #opinions

**Comment by theacodes** [comment]
Oh I forgot to justify this, here's some benefits:

1. Requires a set of eyeballs *and* CI to pass before a release is cut.
2. Allows non-maintainers to initiate a release. Maintainers can approve the release.

**Comment by SethMichaelLarson** [comment]
What if we did away with release branches entirely and just stuck with master? Like you said, any branch can be a release branch as long as on of us reviews it and it passes CI. Is tagging not sufficient?

**Comment by theacodes** [comment]
That's exactly what I'm proposing. I updated my comment to be a little more explicit on that. The `release-MAJOR.MINOR.PATCH` branch is just short-lived branch to initiate the Pull Request.

**Comment by SethMichaelLarson** [comment]
Sorry I misunderstood from the formatting of the release branches! I'm 100% on board having our release process be automated with this model. Eager to hear from others.

**Comment by shazow** [comment]
The thought behind the master/release dichotomy was for people who wanted to pull the library straight from git (as opposed to pypi), "master" would always point to the latest cutting edge work (which might be broken sometimes as things happen, but that's what you'd base your new patches onto) and "release" would always point to the last release (considered to be more "stable", safe to use in production).

By removing the release branch, the user would need to rely on fetching the latest tag by semantic sorting, which is probably fine?

> Oh I forgot to justify this, here's some benefits:
>
> 1. Requires a set of eyeballs *and* CI to pass before a release is cut.
> 2. Allows non-maintainers to initiate a release. Maintainers can approve the release.

Is this different from making a PR into the `release` branch?

I am generally :+1: for automation.

My only concern would be leaving publishing credentials on Travis, and it would mean that release commits would no longer be GPG-signed by the maintainer?

**Comment by theacodes** [comment]
> By removing the release branch, the user would need to rely on fetching the latest tag by semantic sorting, which is probably fine?

I think so. At least, no one has complained in the other libraries I maintain.

> Is this different from making a PR into the release branch?

Nope but avoids confusion and silly errors like #1393.

> My only concern would be leaving publishing credentials on Travis

We can still make this manual, and should probably at least start with it being manual.

> it would mean that release commits would no longer be GPG-signed by the maintainer?

Since the release commit would be the product of merging a PR, GitHub would make it a verified commit.

**Comment by shazow** [comment]
> Nope but avoids confusion and silly errors like #1393.

Hmmm not sure how that would be different, we're still relying on reviewing a diff to catch it either way?

On the other hand, adding a check to the release script would definitely help. :)

> Since the release commit would be the product of merging a PR, GitHub would make it a verified commit.

Is a verified commit different from a signed commit? Or do you mean we'd expect the PR to include a signed commit?

Honestly I'm not sure how important this is in practice. I know of at least one person who cares about such things, but I'm sure the world will keep rotating if urllib3 didn't have signed code releases. Just seems like the right thing to do.

Anyways to be clear: I'm not opposed.

I imagine if I were still doing releases, the release script still sounds like a better workflow for me, but it's totally cool if your process fits better for your own workflow. :)

**Comment by theacodes** [comment]
We absolutely can and should continue to script as much as we can, I just am not currently seeing the benefit of the underlying git manipulations of the present model, and I think this model would make it a bit more transparent and clear.

**Comment by shazow** [comment]
By git manipulation, you mean re-using the release branch instead of making a fresh one every time?

**Comment by shazow** [comment]
One other fun side effect of re-using the release branch (aside from always having a stable branch for consumers to pull from): If you render the git graph for the `release` branch, you get to see all the releases getting merged into it. :)

**Comment by sigmavirus24** [comment]
So it seems like the problem is that git doesn't list tags in chronological order by default. There is a way to do that:`git tag --sort "-*authordate"` will list the most recent tag first, so someone could do `git checkout -b stable $(git tag --sort '-*authordate' | head -1)`. It's also not been entirely clear to me who the target audience is for the `release` branch. I think I must be missing something, because on all the projects I've worked on in the last few years, having a stable release branch like this has never been a request nor is it something that (when I did it with github3.py) was something people noticed, cared about, or appreciated and it only created more work for me. I'm not against keeping it, I'd just like to better understand who is using it, what their reasons are, and if there's a method that's less onerous for maintainers to follow that can satisfy those folks.

**Comment by shazow** [comment]
We have all kinds of tags, and backported fix tags are a thing too--sorting by authordate does not solve the problem unfortunately. (For more fun about extracting semantic versions from tags, tune in to the vgo debates in the Go community.)

Maintainers should definitely not use something that feels onerous.

**To me,** it was more of a matter of reusing a branch (and getting the side effect benefits) vs making and deleting a fresh one every release. It seemed more onerous to not use a release branch. (The rest, whether it's a PR or whether we use a script or whether we make a separate commit for the changelog, that's all equivalent for both scenarios.)

Back in the day I had some things deployed directly from git pulling from the `release` branch. It was convenient, but could have setup a process to check for the latest semantic tag and switch to that instead (also seemed more onerous).

🤷

**Comment by sigmavirus24** [comment]
>  backported fix tags are a thing too--sorting by authordate does not solve the problem unfortunately.

That makes sense. I can't remember, personally, the last time we backported a fix from one release to another older release version and released the backported version after the newer version. (I'm not doubting this has happened before, I just don't remember the project doing it.) That's why I suggested authordate.

**Comment by SethMichaelLarson** [comment]
Hi Daniele, thanks for bringing this to our attention! I wasn't even aware we did signed releases. Will have to investigate that more and bake it into our new release process.

**Comment by eriol** [comment]
Hello Seth, many thanks for the quick reply and for taking care of this! For the 1.23 release I will change the Debian packaging to don't check the signature.
I don't know if this can be useful to you, last release was signed by @Lukasa, I have his key in the Debian package repository:
https://salsa.debian.org/python-team/modules/python-urllib3/blob/debian/master/debian/upstream/signing-key.asc

Thanks again for looking at this! :smile:

**Comment by sigmavirus24** [comment]
@eriol well Cory won't be releasing things anymore. And this is the fundamental problem with people relying on that. Many projects have several people who can release and add more people as the need arises. If @theacodes or @SethMichaelLarson or I had signed it, it would have appeared to be broken or potentially hacked. I don't urllib3 (or really any PyPI uploaded project) should be signing releases. Especially since PyPI really doesn't provide an interface for using that.

**Comment by sigmavirus24** [comment]
Also worth noting that Warehouse/PyPI is working towards obsoleting gpg signatures like this

**Comment by pquentin** [comment]
The relevant Warehouse/PyPI issue: https://github.com/pypa/warehouse/issues/3810

**Comment by sigmavirus24** [comment]
TUF will provide a way for Warehouse to sign artifacts so people can verify aspects of the package. That is the way forward, but it's presently not being actively worked on. This is why GPG signatures haven't been actively removed. Still, I agree with everything that was said in the linked issue from the Warehouse developers.

TUF, however, if it were alive today would not require us to do anything extra.

**Comment by eriol** [comment]
@sigmavirus24 thanks for pointing that, I was aware of it, and sorry if my comment was too strong, take it more as «we rely on it if upstream provide it» :smile: Unfortunately is a very busy moment for me, and I was in a hurry...
Mine was more a question since urllib3 releases was almost always signed[¹] and other project like Django continue doing so[²].
That said, I already removed the sign verification in Debian packaging, and I can understand the point of in pypa/warehouse#3810 although me and Cory are linked through keybase. As I clarified, mine was more a question, not a way to push or force you on a specific procedure.

@theacodes thanks for the release, I will immediately work on uploading 1.24 for Debian.

[¹] https://pypi.debian.net/urllib3
[²] https://pypi.debian.net/django
pypi.debian.net just exposes Warehouse for tools like `uscan`.

**Comment by eriol** [comment]
@theacodes it seems that latest release of requests requires `urllib3 < 1.24`. As far I remember both urllib3 and requests were updated always together. Maybe it's something to take in consideration while discussing the release process:
```
❯ mkvirtualenv --python=/usr/bin/python3 vvv        
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/eriol/.virtualenvs/vvv/bin/python3
Also creating executable in /home/eriol/.virtualenvs/vvv/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
vvv! eriol@mornie ~ ❯ pip install urllib3                       
Collecting urllib3
  Downloading https://files.pythonhosted.org/packages/8c/4b/5cbc4cb46095f369117dcb751821e1bef9dd86a07c968d8757e9204c324c/urllib3-1.24-py2.py3-none-any.whl (117kB)
    100% |████████████████████████████████| 122kB 719kB/s 
Installing collected packages: urllib3
Successfully installed urllib3-1.24
vvv! eriol@mornie ~ ❯ pip install requests
Collecting requests
  Downloading https://files.pythonhosted.org/packages/65/47/7e02164a2a3db50ed6d8a6ab1d6d60b69c4c3fdf57a284257925dfc12bda/requests-2.19.1-py2.py3-none-any.whl (91kB)
    100% |████████████████████████████████| 92kB 541kB/s 
Collecting idna<2.8,>=2.5 (from requests)
  Downloading https://files.pythonhosted.org/packages/4b/2a/0276479a4b3caeb8a8c1af2f8e4355746a97fab05a372e4a2c6a6b876165/idna-2.7-py2.py3-none-any.whl (58kB)
    100% |████████████████████████████████| 61kB 758kB/s 
Collecting certifi>=2017.4.17 (from requests)
  Downloading https://files.pythonhosted.org/packages/56/9d/1d02dd80bc4cd955f98980f28c5ee2200e1209292d5f9e9cc8d030d18655/certifi-2018.10.15-py2.py3-none-any.whl (146kB)
    100% |████████████████████████████████| 153kB 896kB/s 
Collecting urllib3<1.24,>=1.21.1 (from requests)
  Downloading https://files.pythonhosted.org/packages/bd/c9/6fdd990019071a4a32a5e7cb78a1d92c53851ef4f56f62a3486e6a7d8ffb/urllib3-1.23-py2.py3-none-any.whl (133kB)
    100% |████████████████████████████████| 143kB 877kB/s 
Collecting chardet<3.1.0,>=3.0.2 (from requests)
  Downloading https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl (133kB)
    100% |████████████████████████████████| 143kB 884kB/s 
Installing collected packages: idna, certifi, urllib3, chardet, requests
  Found existing installation: urllib3 1.24
    Uninstalling urllib3-1.24:
      Successfully uninstalled urllib3-1.24
Successfully installed certifi-2018.10.15 chardet-3.0.4 idna-2.7 requests-2.19.1 urllib3-1.23
```

**Comment by theacodes** [comment]
This is intentional. We release about a week before requests to give them
time to test and such.

**Comment by eriol** [comment]
Good point! I will package urllib3 1.24 then, but I will wait next week to upload both urllib3 and requests in Debian. Thanks

**Comment by theacodes** [comment]
Sounds good. Thank you!

**Comment by SethMichaelLarson** [comment]
Alright everyone I took a first stab at this by toying around with Travis a lot.
Here's the example repository I built: https://github.com/SethMichaelLarson/python-travis-deploy

The workflow for a release is:
- Any user creates a branch named `release-x.x.x`, commits `CHANGES.rst` and `urllib3/__init__.py`, and opens a PR to our master branch.
- Unit tests run as usual but integration tests also run against this commit.
- Once this PR is approved it'll be merged into `master` any maintainer can tag that commit and push the tagged commit.
- Travis will build the tag and run the PyPI deployment and draft a GitHub release for the tag of the same name.
- The GitHub release can be approved by any maintainer.

Next steps for me:
- Fork the urllib3 repository and integrate this into that fork and see how the deployment goes.
- Write up release procedure documentation for other maintainers and contributors.

What are other maintainers thoughts?
